src/modules/tools/spindle/SoftSerial/BufferedSoftSerial.cpp

```diff
@@ -0,0 +1,144 @@
+/**
+ * @file    BufferedSoftSerial.cpp
+ * @brief   Software Buffer - Extends mbed Serial functionallity adding irq driven TX and RX
+ * @author  sam grove
+ * @version 1.0
+ * @see
+ *
+ * Copyright (c) 2013
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "BufferedSoftSerial.h"
+#include <stdarg.h>
+
+BufferedSoftSerial::BufferedSoftSerial(PinName tx, PinName rx, const char* name)
+    : SoftSerial(tx, rx, name)
+{
+    SoftSerial::attach(this, &BufferedSoftSerial::rxIrq, SoftSerial::RxIrq);
+
+    return;
+}
+
+int BufferedSoftSerial::readable(void)
+{
+    return _rxbuf.size();  // note: look if things are in the buffer
+}
+
+int BufferedSoftSerial::writeable(void)
+{
+    return 1;   // buffer allows overwriting by design, always true
+}
+
+int BufferedSoftSerial::getc(void)
+{
+    char retval;
+    _rxbuf.pop_front(retval);
+    return (int)retval;
+}
+
+int BufferedSoftSerial::putc(int c)
+{
+    _txbuf.push_back((char)c);
+    BufferedSoftSerial::prime();
+
+    return c;
+}
+
+int BufferedSoftSerial::puts(const char *s)
+{
+    const char* ptr = s;
+
+    while(*(ptr) != 0) {
+        _txbuf.push_back(*(ptr++));
+    }
+    _txbuf.push_back('\n');  // done per puts definition
+    BufferedSoftSerial::prime();
+
+    return (ptr - s) + 1;
+}
+
+int BufferedSoftSerial::printf(const char* format, ...)
+{
+    char buf[256] = {0};
+    int r = 0;
+
+    va_list arg;
+    va_start(arg, format);
+    r = vsprintf(buf, format, arg);
+    // this may not hit the heap but should alert the user anyways
+    if(r > (int) sizeof(buf)) {
+        error("%s %d buffer overwrite!\n", __FILE__, __LINE__);

```

Comment:
do not use error as it is not handled properly. I'm not even sure what it would do.